### PR TITLE
RELATED: RAIL-3560 - Expand ObjRefMap to allow storing any object that has `ref`

### PR DESCRIPTION
**Pull request by lupko**
-  ObjRefMap can now handle cases when object being inserted has id, uri or both
-  the new factory function `newMapForObjectWithRef` allows simple creation of maps for arbitrary objects
   that contain just `ref` property. Such map will extract the essentials from the ref and is ready for
   refs that are both uriRef and identifierRef
-  Testing what insights are available was not correctly checking refs

JIRA: RAIL-3560

---

Supported PR commands:

| Command                | Description            |
| ---------------------- | ---------------------- |
| `ok to test`           | Re-run standard checks |
| `extended test`        | BackstopJS tests       |
| `extended check sonar` | SonarQube tests        |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/4515/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6130/ : FAILURE in 5m 36s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/4516/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6131/ : SUCCESS in 14m 24s

File: libs/sdk-ui-dashboard/src/_staging/metadata/objRefMap.ts [hunk 1]
```diff
@@ -80,28 +91,60 @@ export class ObjRefMap<T> {
...
+        if (!this.shouldInsert(uri, identifier)) {
+            return;
+        }
+
+        if (uri) {
+            this.itemsByUri[uri] = item;
+        }
+
+        if (identifier) {
+            this.itemsByIdentifier[`${this.idRefToKey(identifier, this.config.type)}`] = item;
```
Comment (no23reason) [inline libs/sdk-ui-dashboard/src/_staging/metadata/objRefMap.ts:136]: Detail: could be just
```ts
this.itemsByIdentifier[this.idRefToKey(identifier, this.config.type)] = item;
```
right? Or why the template string?

Comment (lupko) [reply libs/sdk-ui-dashboard/src/_staging/metadata/objRefMap.ts]: fixed

**Review by no23reason — APPROVED** [review]
LGTM

File: libs/sdk-ui-dashboard/src/_staging/metadata/tests/objRefMap.test.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        });
+
+        it("should allow lookup only by uri if the stored object only has uri", () => {
+            const map = newMapForObjectWithRef(MixedTestData);
+
+            expect(map.get(idRef("id2"))).toBeUndefined();
+            expect(map.get(uriRef("uri2"))).toBeDefined();
+        });
+
+        it("should bomb if trying to insert object that has both id&uri but previous item with same id and no uri was inserted ", () => {
```
Comment (no23reason) [inline libs/sdk-ui-dashboard/src/_staging/metadata/tests/objRefMap.test.ts:61]: Should this really bomb? What if I insert object with just id and then the same object with both id and uri? That sounds ok to me.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR1556

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/1285/ : SUCCESS in 1m 25s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/1561/ : SUCCESS in 17s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/331/ : SUCCESS in 14m 28s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6140/ : SUCCESS in 14m 04s

**Pull request merged**